In [20]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.optimizers import SGD
import matplotlib.pyplot as plt

### Split and Scale the Fasion MNIST Data into Train and Test Sets

In [2]:
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()


input_dim = 28 * 28 # pixel length and width of each image
output_dim = nb_classes = 10
nb_epoch = 20

x_train = x_train.reshape(60000, input_dim)
x_test = x_test.reshape(10000, input_dim)
x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255 # normalize by dividing by maximum pixel value
x_test /= 255

4423680/4422102 [==============================] - 10s 2us/step


In [3]:
y_train = to_categorical(y_train, nb_classes)
y_test = to_categorical(y_test, nb_classes)

### Define the Parameters to be Tuned for Each Deep Learning Model

In [30]:
def fit_model(x_train=x_train, x_test=x_test, y_train=y_train, y_test=y_test,
              activation='relu', optimizer='sgd', loss='categorical_crossentropy', batch_size=128):
    
    model = Sequential()
    model.add(Dense(1028, input_shape=x_train[0].shape, activation=activation))
    model.add(Dense(1024, activation=activation))
    model.add(Dense(512, activation=activation))
    model.add(Dense(256, activation=activation))
    model.add(Dense(128, activation=activation))
    model.add(Dense(64, activation=activation))
    model.add(Dense(10, activation='softmax'))
    model.compile(optimizer=optimizer, loss=loss, metrics=['accuracy'])
    history = model.fit(x_train, y_train, batch_size=batch_size, epochs=15, verbose=1, validation_data=(x_test, y_test))
    parameters = {'activation':activation, 'optimizer':optimizer, 'loss':loss, 'batch_size':batch_size}
    return history, parameters

In [31]:
activations = ['relu', 'tanh', 'sigmoid']
optimizers = ['sgd', 'adam']
losses = ['categorical_crossentropy', 'hinge']
batches = [64, 128, 256]

### Fit the Models and Organize the Data

In [32]:
history_list = []
parameters_list = []
for a in activations:
    for o in optimizers:
        for l in losses:
            for b in batches:
                history, parameters = fit_model(activation=a, optimizer=o, loss=l, batch_size=b)
                history_list.append(history)
                parameters_list.append(parameters)

Train on 60000 samples, validate on 10000 samples
Epoch 1/15
60000/60000 [==============================] - 4s 68us/sample - loss: 0.8849 - accuracy: 0.7065 - val_loss: 0.5650 - val_accuracy: 0.7996
Epoch 2/15
60000/60000 [==============================] - 4s 61us/sample - loss: 0.4960 - accuracy: 0.8255 - val_loss: 0.4818 - val_accuracy: 0.8258
Epoch 3/15
60000/60000 [==============================] - 4s 62us/sample - loss: 0.4318 - accuracy: 0.8473 - val_loss: 0.4408 - val_accuracy: 0.8412
Epoch 4/15
60000/60000 [==============================] - 4s 60us/sample - loss: 0.3976 - accuracy: 0.8597 - val_loss: 0.4557 - val_accuracy: 0.8290
Epoch 5/15
60000/60000 [==============================] - 4s 60us/sample - loss: 0.3717 - accuracy: 0.8672 - val_loss: 0.4071 - val_accuracy: 0.8523
Epoch 6/15
60000/60000 [==============================] - 4s 61us/sample - loss: 0.3512 - accuracy: 0.8730 - val_loss: 0.3868 - val_accuracy: 0.8577
Epoch 7/15
60000/60000 [==============================] 

60000/60000 [==============================] - 1s 23us/sample - loss: 0.4333 - accuracy: 0.8479 - val_loss: 0.5317 - val_accuracy: 0.8004
Epoch 10/15
60000/60000 [==============================] - 1s 23us/sample - loss: 0.4248 - accuracy: 0.8501 - val_loss: 0.6334 - val_accuracy: 0.7886
Epoch 11/15
60000/60000 [==============================] - 2s 25us/sample - loss: 0.4090 - accuracy: 0.8564 - val_loss: 0.4648 - val_accuracy: 0.8333
Epoch 12/15
60000/60000 [==============================] - 1s 22us/sample - loss: 0.3989 - accuracy: 0.8593 - val_loss: 0.4323 - val_accuracy: 0.8467
Epoch 13/15
60000/60000 [==============================] - 1s 23us/sample - loss: 0.3940 - accuracy: 0.8615 - val_loss: 0.4302 - val_accuracy: 0.8451
Epoch 14/15
60000/60000 [==============================] - 1s 22us/sample - loss: 0.3865 - accuracy: 0.8640 - val_loss: 0.4498 - val_accuracy: 0.8427
Epoch 15/15
60000/60000 [==============================] - 2s 25us/sample - loss: 0.3782 - accuracy: 0.8656 - va

60000/60000 [==============================] - 2s 31us/sample - loss: 1.0797 - accuracy: 0.1230 - val_loss: 1.0793 - val_accuracy: 0.1429
Epoch 2/15
60000/60000 [==============================] - 1s 24us/sample - loss: 1.0790 - accuracy: 0.1594 - val_loss: 1.0786 - val_accuracy: 0.1722
Epoch 3/15
60000/60000 [==============================] - 1s 24us/sample - loss: 1.0783 - accuracy: 0.1896 - val_loss: 1.0779 - val_accuracy: 0.2001
Epoch 4/15
60000/60000 [==============================] - 1s 24us/sample - loss: 1.0775 - accuracy: 0.2133 - val_loss: 1.0771 - val_accuracy: 0.2228
Epoch 5/15
60000/60000 [==============================] - 1s 23us/sample - loss: 1.0767 - accuracy: 0.2342 - val_loss: 1.0763 - val_accuracy: 0.2447
Epoch 6/15
60000/60000 [==============================] - 1s 24us/sample - loss: 1.0758 - accuracy: 0.2554 - val_loss: 1.0753 - val_accuracy: 0.2600
Epoch 7/15
60000/60000 [==============================] - 1s 24us/sample - loss: 1.0746 - accuracy: 0.2703 - val_loss

Epoch 15/15
60000/60000 [==============================] - 3s 46us/sample - loss: 0.1875 - accuracy: 0.9291 - val_loss: 0.3287 - val_accuracy: 0.8919
Train on 60000 samples, validate on 10000 samples
Epoch 1/15
60000/60000 [==============================] - 2s 37us/sample - loss: 0.5493 - accuracy: 0.8033 - val_loss: 0.4451 - val_accuracy: 0.8336
Epoch 2/15
60000/60000 [==============================] - 2s 28us/sample - loss: 0.3650 - accuracy: 0.8659 - val_loss: 0.4203 - val_accuracy: 0.8495
Epoch 3/15
60000/60000 [==============================] - 2s 29us/sample - loss: 0.3252 - accuracy: 0.8789 - val_loss: 0.3633 - val_accuracy: 0.8697
Epoch 4/15
60000/60000 [==============================] - 2s 28us/sample - loss: 0.3011 - accuracy: 0.8882 - val_loss: 0.3779 - val_accuracy: 0.8596
Epoch 5/15
60000/60000 [==============================] - 2s 28us/sample - loss: 0.2848 - accuracy: 0.8941 - val_loss: 0.3695 - val_accuracy: 0.8687
Epoch 6/15
60000/60000 [==============================]

60000/60000 [==============================] - 3s 48us/sample - loss: 1.0602 - accuracy: 0.1991 - val_loss: 1.0602 - val_accuracy: 0.1988
Epoch 9/15
60000/60000 [==============================] - 3s 48us/sample - loss: 1.0602 - accuracy: 0.1991 - val_loss: 1.0602 - val_accuracy: 0.1988
Epoch 10/15
60000/60000 [==============================] - 3s 48us/sample - loss: 1.0602 - accuracy: 0.1991 - val_loss: 1.0602 - val_accuracy: 0.1988
Epoch 11/15
60000/60000 [==============================] - 3s 48us/sample - loss: 1.0602 - accuracy: 0.1991 - val_loss: 1.0602 - val_accuracy: 0.1988
Epoch 12/15
60000/60000 [==============================] - 3s 48us/sample - loss: 1.0602 - accuracy: 0.1991 - val_loss: 1.0602 - val_accuracy: 0.1988
Epoch 13/15
60000/60000 [==============================] - 3s 47us/sample - loss: 1.0602 - accuracy: 0.1991 - val_loss: 1.0602 - val_accuracy: 0.1988
Epoch 14/15
60000/60000 [==============================] - 3s 48us/sample - loss: 1.0602 - accuracy: 0.1991 - val

60000/60000 [==============================] - 3s 49us/sample - loss: 0.7614 - accuracy: 0.7668 - val_loss: 0.5595 - val_accuracy: 0.8120
Epoch 2/15
60000/60000 [==============================] - 2s 41us/sample - loss: 0.4999 - accuracy: 0.8289 - val_loss: 0.5225 - val_accuracy: 0.8117
Epoch 3/15
60000/60000 [==============================] - 2s 40us/sample - loss: 0.4511 - accuracy: 0.8433 - val_loss: 0.4642 - val_accuracy: 0.8327
Epoch 4/15
60000/60000 [==============================] - 2s 39us/sample - loss: 0.4232 - accuracy: 0.8508 - val_loss: 0.4519 - val_accuracy: 0.8379
Epoch 5/15
60000/60000 [==============================] - 2s 39us/sample - loss: 0.4048 - accuracy: 0.8581 - val_loss: 0.4317 - val_accuracy: 0.8452
Epoch 6/15
60000/60000 [==============================] - 2s 39us/sample - loss: 0.3905 - accuracy: 0.8614 - val_loss: 0.4401 - val_accuracy: 0.8422
Epoch 7/15
60000/60000 [==============================] - 2s 39us/sample - loss: 0.3789 - accuracy: 0.8663 - val_loss

Epoch 15/15
60000/60000 [==============================] - 4s 74us/sample - loss: 0.9445 - accuracy: 0.8000 - val_loss: 0.9456 - val_accuracy: 0.7911
Train on 60000 samples, validate on 10000 samples
Epoch 1/15
60000/60000 [==============================] - 3s 48us/sample - loss: 1.0546 - accuracy: 0.3491 - val_loss: 1.0327 - val_accuracy: 0.4671
Epoch 2/15
60000/60000 [==============================] - 2s 40us/sample - loss: 1.0194 - accuracy: 0.5309 - val_loss: 1.0101 - val_accuracy: 0.5423
Epoch 3/15
60000/60000 [==============================] - 2s 40us/sample - loss: 1.0052 - accuracy: 0.5444 - val_loss: 1.0013 - val_accuracy: 0.5478
Epoch 4/15
60000/60000 [==============================] - 2s 41us/sample - loss: 0.9953 - accuracy: 0.6004 - val_loss: 0.9904 - val_accuracy: 0.6230
Epoch 5/15
60000/60000 [==============================] - 2s 40us/sample - loss: 0.9873 - accuracy: 0.6261 - val_loss: 0.9854 - val_accuracy: 0.6260
Epoch 6/15
60000/60000 [==============================]

60000/60000 [==============================] - 5s 84us/sample - loss: 0.3284 - accuracy: 0.8795 - val_loss: 0.3846 - val_accuracy: 0.8642
Epoch 9/15
60000/60000 [==============================] - 5s 84us/sample - loss: 0.3203 - accuracy: 0.8821 - val_loss: 0.3893 - val_accuracy: 0.8570
Epoch 10/15
60000/60000 [==============================] - 5s 83us/sample - loss: 0.3170 - accuracy: 0.8832 - val_loss: 0.4464 - val_accuracy: 0.8454
Epoch 11/15
60000/60000 [==============================] - 5s 85us/sample - loss: 0.3139 - accuracy: 0.8842 - val_loss: 0.3710 - val_accuracy: 0.8670
Epoch 12/15
60000/60000 [==============================] - 5s 84us/sample - loss: 0.3104 - accuracy: 0.8855 - val_loss: 0.3593 - val_accuracy: 0.8696
Epoch 13/15
60000/60000 [==============================] - 5s 83us/sample - loss: 0.3062 - accuracy: 0.8856 - val_loss: 0.4053 - val_accuracy: 0.8553
Epoch 14/15
60000/60000 [==============================] - 5s 83us/sample - loss: 0.3048 - accuracy: 0.8874 - val

60000/60000 [==============================] - 6s 104us/sample - loss: 0.9612 - accuracy: 0.6989 - val_loss: 0.9496 - val_accuracy: 0.7532
Epoch 2/15
60000/60000 [==============================] - 6s 92us/sample - loss: 0.9500 - accuracy: 0.7510 - val_loss: 0.9436 - val_accuracy: 0.7821
Epoch 3/15
60000/60000 [==============================] - 7s 114us/sample - loss: 0.9466 - accuracy: 0.7678 - val_loss: 0.9428 - val_accuracy: 0.7872
Epoch 4/15
60000/60000 [==============================] - 5s 90us/sample - loss: 0.9421 - accuracy: 0.7901 - val_loss: 0.9449 - val_accuracy: 0.7755
Epoch 5/15
60000/60000 [==============================] - 5s 91us/sample - loss: 0.9427 - accuracy: 0.7870 - val_loss: 0.9422 - val_accuracy: 0.7892
Epoch 6/15
60000/60000 [==============================] - 5s 91us/sample - loss: 0.9412 - accuracy: 0.7946 - val_loss: 0.9431 - val_accuracy: 0.7846
Epoch 7/15
60000/60000 [==============================] - 5s 90us/sample - loss: 0.9409 - accuracy: 0.7957 - val_lo

Epoch 15/15
60000/60000 [==============================] - 2s 29us/sample - loss: 0.9386 - accuracy: 0.8074 - val_loss: 0.9398 - val_accuracy: 0.8009
Train on 60000 samples, validate on 10000 samples
Epoch 1/15
60000/60000 [==============================] - 5s 79us/sample - loss: 2.3064 - accuracy: 0.0999 - val_loss: 2.3029 - val_accuracy: 0.1000
Epoch 2/15
60000/60000 [==============================] - 4s 68us/sample - loss: 2.3033 - accuracy: 0.0987 - val_loss: 2.3032 - val_accuracy: 0.1000
Epoch 3/15
60000/60000 [==============================] - 4s 68us/sample - loss: 2.3035 - accuracy: 0.0990 - val_loss: 2.3033 - val_accuracy: 0.1000
Epoch 4/15
60000/60000 [==============================] - 4s 68us/sample - loss: 2.3033 - accuracy: 0.0984 - val_loss: 2.3034 - val_accuracy: 0.1000
Epoch 5/15
60000/60000 [==============================] - 4s 68us/sample - loss: 2.3034 - accuracy: 0.1000 - val_loss: 2.3033 - val_accuracy: 0.1000
Epoch 6/15
60000/60000 [==============================]

60000/60000 [==============================] - 1s 25us/sample - loss: 2.3028 - accuracy: 0.1007 - val_loss: 2.3027 - val_accuracy: 0.1000
Epoch 9/15
60000/60000 [==============================] - 1s 25us/sample - loss: 2.3028 - accuracy: 0.0994 - val_loss: 2.3027 - val_accuracy: 0.1000
Epoch 10/15
60000/60000 [==============================] - 1s 25us/sample - loss: 2.3029 - accuracy: 0.0992 - val_loss: 2.3026 - val_accuracy: 0.1000
Epoch 11/15
60000/60000 [==============================] - 2s 25us/sample - loss: 2.3028 - accuracy: 0.1007 - val_loss: 2.3026 - val_accuracy: 0.1000
Epoch 12/15
60000/60000 [==============================] - 1s 25us/sample - loss: 2.3028 - accuracy: 0.1001 - val_loss: 2.3027 - val_accuracy: 0.1000
Epoch 13/15
60000/60000 [==============================] - 1s 25us/sample - loss: 2.3028 - accuracy: 0.0998 - val_loss: 2.3026 - val_accuracy: 0.1000
Epoch 14/15
60000/60000 [==============================] - 2s 25us/sample - loss: 2.3028 - accuracy: 0.0972 - val

60000/60000 [==============================] - 2s 34us/sample - loss: 1.0800 - accuracy: 0.1000 - val_loss: 1.0800 - val_accuracy: 0.1000
Epoch 2/15
60000/60000 [==============================] - 2s 26us/sample - loss: 1.0800 - accuracy: 0.1000 - val_loss: 1.0800 - val_accuracy: 0.1000
Epoch 3/15
60000/60000 [==============================] - 2s 25us/sample - loss: 1.0800 - accuracy: 0.1000 - val_loss: 1.0800 - val_accuracy: 0.1000
Epoch 4/15
60000/60000 [==============================] - 2s 25us/sample - loss: 1.0800 - accuracy: 0.1000 - val_loss: 1.0800 - val_accuracy: 0.1000
Epoch 5/15
60000/60000 [==============================] - 2s 26us/sample - loss: 1.0800 - accuracy: 0.1000 - val_loss: 1.0800 - val_accuracy: 0.1000
Epoch 6/15
60000/60000 [==============================] - 2s 26us/sample - loss: 1.0800 - accuracy: 0.1000 - val_loss: 1.0800 - val_accuracy: 0.1000
Epoch 7/15
60000/60000 [==============================] - 2s 26us/sample - loss: 1.0800 - accuracy: 0.1000 - val_loss

Epoch 15/15
60000/60000 [==============================] - 3s 47us/sample - loss: 0.2363 - accuracy: 0.9131 - val_loss: 0.3560 - val_accuracy: 0.8833
Train on 60000 samples, validate on 10000 samples
Epoch 1/15
60000/60000 [==============================] - 2s 39us/sample - loss: 1.6954 - accuracy: 0.3193 - val_loss: 1.0646 - val_accuracy: 0.5752
Epoch 2/15
60000/60000 [==============================] - 2s 29us/sample - loss: 0.7820 - accuracy: 0.7184 - val_loss: 0.6398 - val_accuracy: 0.7797
Epoch 3/15
60000/60000 [==============================] - 2s 29us/sample - loss: 0.5655 - accuracy: 0.7978 - val_loss: 0.5394 - val_accuracy: 0.8043
Epoch 4/15
60000/60000 [==============================] - 2s 29us/sample - loss: 0.4873 - accuracy: 0.8177 - val_loss: 0.5004 - val_accuracy: 0.8068
Epoch 5/15
60000/60000 [==============================] - 2s 29us/sample - loss: 0.4482 - accuracy: 0.8325 - val_loss: 0.4609 - val_accuracy: 0.8316
Epoch 6/15
60000/60000 [==============================]

60000/60000 [==============================] - 3s 50us/sample - loss: 1.0601 - accuracy: 0.1998 - val_loss: 1.0600 - val_accuracy: 0.1999
Epoch 9/15
60000/60000 [==============================] - 3s 50us/sample - loss: 1.0600 - accuracy: 0.1998 - val_loss: 1.0600 - val_accuracy: 0.1999
Epoch 10/15
60000/60000 [==============================] - 3s 49us/sample - loss: 1.0602 - accuracy: 0.1992 - val_loss: 1.0600 - val_accuracy: 0.1999
Epoch 11/15
60000/60000 [==============================] - 3s 50us/sample - loss: 1.0600 - accuracy: 0.1999 - val_loss: 1.0600 - val_accuracy: 0.1999
Epoch 12/15
60000/60000 [==============================] - 3s 50us/sample - loss: 1.0600 - accuracy: 0.1999 - val_loss: 1.0600 - val_accuracy: 0.1999
Epoch 13/15
60000/60000 [==============================] - 3s 50us/sample - loss: 1.0600 - accuracy: 0.1999 - val_loss: 1.0600 - val_accuracy: 0.1999
Epoch 14/15
60000/60000 [==============================] - 3s 50us/sample - loss: 1.0600 - accuracy: 0.1999 - val

In [34]:
d_list1 = []
d_list2 = []
for i in range(len(parameters_list)):
    dict1 = {'accuracy':history_list[i].history['accuracy'][-1],
                'val_accuracy':history_list[i].history['val_accuracy'][-1],
                'loss':history_list[i].history['loss'][-1],
                'val_loss':history_list[i].history['val_loss'][-1]}
    d_list1.append(dict1)
    d_list2.append(parameters_list[i])
    
df1 =  pd.DataFrame(d_list1)
df2 = pd.DataFrame(d_list2)

df3 = pd.concat([df1, df2], axis=1)

In [38]:
df3['diff_accuracies_%'] = 100 * abs(df3['accuracy'] - df3['val_accuracy']) / df3['val_accuracy']
df3.sort_values('val_accuracy', ascending=False, inplace=True)
df3.head(10)

accuracy  val_accuracy      loss  val_loss activation optimizer  \
7   0.929100        0.8919  0.187482  0.328689       relu      adam   
8   0.927300        0.8880  0.188288  0.358728       relu      adam   
6   0.924117        0.8872  0.199292  0.367700       relu      adam   
20  0.916033        0.8862  0.222674  0.324572       tanh      adam   
30  0.916350        0.8842  0.226671  0.348607    sigmoid      adam   
31  0.913083        0.8833  0.236333  0.355980    sigmoid      adam   
32  0.901383        0.8807  0.267974  0.347163    sigmoid      adam   
0   0.907017        0.8774  0.248732  0.339000       relu       sgd   
19  0.903733        0.8771  0.256810  0.348081       tanh      adam   
12  0.895833        0.8758  0.284454  0.344295       tanh       sgd   

                        loss  batch_size  diff_accuracies_%  
7   categorical_crossentropy         128           4.170868  
8   categorical_crossentropy         256           4.425672  
6   categorical_crossentropy          64           4.161032  
20  categorical_crossentropy         256           3.366432  
30  categorical_crossentropy          64           3.636059  
31  categorical_crossentropy         128           3.371823  
32  categorical_crossentropy         256           2.348512  
0   categorical_crossentropy          64           3.375509  
19  categorical_crossentropy         128           3.036521  
12  categorical_crossentropy          64           2.287429

From the above results, it appears the specific combination of: activation=relu, optimizer=adam, loss=categorical_crossentropy was the strongest parameter tunings for the first 3 models with a batch size of 128 being the best tuning out of those models. If you wanted to sacrifice accuracy a hair for slightly better confidence that your models werent overfit, the model with activation=tanh and batch_size=256 was the next best performer with ~88.6% accuracy. For comparison, the best model only had an accuracy of ~89.2%, so the differences between model performance is pretty negligible as they are both fairly accurate.